File: java/src/DataStructures/Queue.java
package DataStructures;

/**
 * Queue - FIFO queue
 * 		 - Linked List implementation
 * 		 - constant time enqueue() and dequeue() methods
 *
 * @param <Type>
 */
public class Queue<Type> {
	
	private Node<Type> head;
	private Node<Type> tail;
	private int size;
	
	private class Node<Type> {
		private Type data;
		private Node next;
		public Node(Type data, Node next) {
			this.data = data;
			this.next = next;
		}
	}
	
	public Queue() {
		head = null;
		tail = null;
		size = 0;
	}
	
	/**
	 * enqueue(data)
	 * Insert item at the end of the list
	 * @param data
	 */
	public void enqueue(Type data) {
		if (isEmpty()) {
			this.head = new Node(data, null);
			this.tail = this.head;
		}
		else {
			Node oldtail = this.tail;
			this.tail = new Node(data, null);
			oldtail.next = tail;
		}
		size ++;
	}
	
	/**
	 * dequeue()
	 * Remove item from the beginning of the list
	 * @return Node
	 */
	public Node dequeue() {
		if(isEmpty()) return null;
		Node oldhead = this.head;
		this.head = oldhead.next;
		if(isEmpty()) this.tail = null;
		size--;
		return oldhead;
	}
	
	/**
	 * isEmpty()
	 * @return
	 */
	public Boolean isEmpty() {
		return head == null;
	}
	
	public String toString() {
		String[] toPrint = new String[size];
		Node tmp = this.head;
		int i = 0;
		while (tmp != null) {
			toPrint[i] = (String) tmp.data;
			tmp = tmp.next;
			i++;
		}
		return String.join(" ", toPrint);
	}
	
	public static void main(String[] args) {
		Queue<String> q = new Queue<String>();
		String input = "a b c d e f g h i j k l m n o p";
		for (String str: input.split(" ")) {
			q.enqueue(str);
		}
		System.out.println(q);
		q.dequeue();
		q.dequeue();
		System.out.println(q);
		
	}
	
}

File: python/DataStructures/Queue.py
"""

Queue implementation (first-in-first-out) uing LinkedList

enqueue(key)	| O(1)
dequeue()		| O(1)
is_empty()		| O(1)

"""

class Node:
	def __init__(self, data, next=None):
		self.data  = data 
		self.next = next

class Queue:
	def __init__(self):
		self.head = None
		self.tail = None

	def enqueue(self, data):
		if self.is_empty():
			self.head = Node(data)
			self.tail = self.head
		else:
			oldtail = self.tail
			self.tail = Node(data)
			oldtail.next = self.tail

	def dequeue(self):
		if self.is_empty():
			return None

		oldhead = self.head
		self.head = oldhead.next
		if self.is_empty():
			self.tail = None
		return oldhead

	def is_empty(self):
		return self.head == None

	def __str__(self):
		"""
		Prints the current list in the form of a Python list            
		"""
		current = self.head
		toPrint = []
		while current != None:
			toPrint.append(current.data)
			current = current.next
		return str(toPrint) 

if __name__ == "__main__":
	s = Queue()
	for i in range(0,20):
		s.enqueue(i)
	s.dequeue()
	s.dequeue()
	print(s)